# Lazizbek, Math 548, HW #6

#Homework-6

**Solve the following systems of equations first using the Jacobi iteration method and then, using the Gauss-Seidel iteration method. In each case, answer these questions.**


*   Does the Jacobi iteration converge to a solution? If yes, what is the solution? If no, what is the justification?
*   Does the Gauss-Seidel iteration converge to a solution? If yes, what is the solution? If no, what is the justification?



1.   4x1 + x2 = 5, x1 - 5x2 = -4

2.   [ 2 -1  0][x1]   [ 1]

     [-1  2 -1][x2] = [ 0]

     [ 0 -1  2][x3]   [ 1]


3.   [ 2 -1  1][x1]   [-1]
    
     [ 2  2  2][x2] = [ 4]
    
     [-1 -1  2][x3]   [ 5]


4.   [ 1  2 -2][x1]   [ 7]

     [ 1  1  1][x2] = [ 2]

     [ 2  2  1][x3]   [ 5]



# Finding eigenstuff of a matrix
[Source:](https://pythonnumericalmethods.berkeley.edu/notebooks/chapter15.04-Eigenvalues-and-Eigenvectors-in-Python.html)

In [10]:
import numpy as np
from numpy.linalg import eig
a = np.array([[0, 2],
              [2, 3]])
w,v=eig(a)
print('E-value:', w)
print('E-vector', v)

E-value: [-1.  4.]
E-vector [[-0.89442719 -0.4472136 ]
 [ 0.4472136  -0.89442719]]


# Solving Ax=b matrix equation
[Source:](https://stackoverflow.com/questions/22163113/matrix-multiplication-solve-ax-b-solve-for-x)

# Problem 1

In [11]:
# To solve Ax=b, We start by constructing the arrays for A and b.

A = np.array([[ 4,  1],
              [ 1, -5]])
b = np.transpose(np.array([ 5, -4]))
# To solve the system we do

x = np.linalg.solve(A,b)
print("Real solution: ")
print(x)

Real solution: 
[1. 1.]


**Using Jacobi Iteration:**

In [12]:
A = np.array([[ 4,  1],
              [ 1, -5]])

L = np.array([[ 0,  0],
              [ 1,  0]])

D = np.array([[ 4,  0],
              [ 0,  -5]])

U = np.array([[ 0,  1],
              [ 0,  0]])

D_inverse =  np. linalg. inv(D)
b = np.transpose(np.array([ 5, -4]))
D_inverse_b = np.dot(D_inverse, b)

BJ = np.dot(-D_inverse, L+U)
w,v=eig(BJ)

print('BJ evalues:', w)

x0 = np.transpose(np.array([0,0]))
x = list();
x.append(x0)

for i in range(10):
  x1 = np.dot(BJ, x0)+ D_inverse_b
  x0 = x1
  x.append(x1)
Aproximations = np.array(x)
print(Aproximations)

BJ evalues: [0.+0.2236068j 0.-0.2236068j]
[[0.         0.        ]
 [1.25       0.8       ]
 [1.05       1.05      ]
 [0.9875     1.01      ]
 [0.9975     0.9975    ]
 [1.000625   0.9995    ]
 [1.000125   1.000125  ]
 [0.99996875 1.000025  ]
 [0.99999375 0.99999375]
 [1.00000156 0.99999875]
 [1.00000031 1.00000031]]


*Because the spectral radius of matrix BJ,  P(BJ) = 0.22 < 1, Jacobi Iteration does converge.*

**Using Gauss-Siedell Iteration:**

In [13]:
A = np.array([[ 4,  1],
              [ 1, -5]])

L = np.array([[ 0,  0],
              [ 1,  0]])

D = np.array([[ 4,  0],
              [ 0,  -5]])

U = np.array([[ 0,  1],
              [ 0,  0]])

DL_inverse =  np. linalg. inv(D+L)
b = np.transpose(np.array([ 5, -4]))
DL_inverse_b = np.dot(DL_inverse, b)

BGS = np.dot(-DL_inverse, U)
w,v=eig(BGS)

print('BGS evalues:', w)

x0 = np.transpose(np.array([ 0, 0]))
x = list();
x.append(x0)

for i in range(10):
  x1 = np.dot(BGS, x0)+ DL_inverse_b
  x0 = x1
  x.append(x1)
Aproximations = np.array(x)
print(Aproximations)

BGS evalues: [ 0.   -0.05]
[[0.         0.        ]
 [1.25       1.05      ]
 [0.9875     0.9975    ]
 [1.000625   1.000125  ]
 [0.99996875 0.99999375]
 [1.00000156 1.00000031]
 [0.99999992 0.99999998]
 [1.         1.        ]
 [1.         1.        ]
 [1.         1.        ]
 [1.         1.        ]]


*Because the spectral radius of matrix BGS, P(BGS) = 0.05 < 1, Gauss-Siedell Iteration does converge.*

**Comparison of BJ vs BGS, P(BJ) =  0.22 < 0.05 = P(BGS), Jacobi Iteration converges slower than Gauss-Siedell Iteration!**

#Problem 2

In [14]:
# To solve Ax=b, We start by constructing the arrays for A and b.

A = np.array([[ 2, -1,  0],
              [-1,  2, -1],
              [ 0, -1,  2]])
b = np.transpose(np.array([ 1,  0,  1]))
# To solve the system we do

x = np.linalg.solve(A,b)
print("Real solution: ")
print(x)

Real solution: 
[1. 1. 1.]


**Using Jacobi Iteration:**

In [17]:
A = np.array([[ 2, -1,  0],
              [-1,  2, -1],
              [ 0, -1,  2]])

L = np.array([[ 0,  0,  0],
              [-1,  0,  0],
              [ 0, -1,  0]])

D = np.array([[ 2,  0,  0],
              [ 0,  2,  0],
              [ 0,  0,  2]])

U = np.array([[ 0, -1,  0],
              [ 0,  0, -1],
              [ 0,  0 ,  0]])

D_inverse =  np. linalg. inv(D)
b = np.transpose(np.array([ 1,  0,  1]))
D_inverse_b = np.dot(D_inverse, b)

BJ = np.dot(-D_inverse, L+U)
w,v=eig(BJ)

print('BJ evalues:', w)

x0 = np.transpose(np.array([ 0,  0,  0]))
x = list();
x.append(x0)

for i in range(10):
  x1 = np.dot(BJ, x0)+ D_inverse_b
  x0 = x1
  x.append(x1)
Aproximations = np.array(x)
print(Aproximations)

BJ evalues: [-7.07106781e-01  4.88975180e-17  7.07106781e-01]
[[0.      0.      0.     ]
 [0.5     0.      0.5    ]
 [0.5     0.5     0.5    ]
 [0.75    0.5     0.75   ]
 [0.75    0.75    0.75   ]
 [0.875   0.75    0.875  ]
 [0.875   0.875   0.875  ]
 [0.9375  0.875   0.9375 ]
 [0.9375  0.9375  0.9375 ]
 [0.96875 0.9375  0.96875]
 [0.96875 0.96875 0.96875]]


*Because the spectral radius of matrix BJ,  P(BJ) = 7.07106781e-01 = 0.7 < 1, Jacobi Iteration does converge.*

**Using Gauss_Siedell Iteration:**

In [18]:
A = np.array([[ 2, -1,  0],
              [-1,  2, -1],
              [ 0, -1,  2]])

L = np.array([[ 0,  0,  0],
              [-1,  0,  0],
              [ 0, -1,  0]])

D = np.array([[ 2,  0,  0],
              [ 0,  2,  0],
              [ 0,  0,  2]])

U = np.array([[ 0, -1,  0],
              [ 0,  0, -1],
              [ 0,  0 ,  0]])

DL_inverse =  np. linalg. inv(D+L)
b = np.transpose(np.array([ 1,  0,  1]))
DL_inverse_b = np.dot(DL_inverse, b)

BGS = np.dot(-DL_inverse, U)
w,v=eig(BGS)

print('BGS evalues:', w)

x0 = np.transpose(np.array([ 0, 0, 0]))
x = list();
x.append(x0)

for i in range(10):
  x1 = np.dot(BGS, x0)+ DL_inverse_b
  x0 = x1
  x.append(x1)
Aproximations = np.array(x)
print(Aproximations)

BGS evalues: [0.00000000e+00 5.00000000e-01 5.55111512e-17]
[[0.         0.         0.        ]
 [0.5        0.25       0.625     ]
 [0.625      0.625      0.8125    ]
 [0.8125     0.8125     0.90625   ]
 [0.90625    0.90625    0.953125  ]
 [0.953125   0.953125   0.9765625 ]
 [0.9765625  0.9765625  0.98828125]
 [0.98828125 0.98828125 0.99414062]
 [0.99414062 0.99414062 0.99707031]
 [0.99707031 0.99707031 0.99853516]
 [0.99853516 0.99853516 0.99926758]]


*Because the spectral radius of matrix BGS,  P(BGS) = 5.00000000e-01 = 0.5 < 1, Gauss-Siedell Iteration does converge.*

**Comparison of BJ vs BGS, P(BJ) =  0.7 < 0.5 = P(BGS), Jacobi Iteration converges slower than Gauss-Siedell Iteration!**

# Problem 3

In [19]:
# To solve Ax=b, We start by constructing the arrays for A and b.

A = np.array([[ 2, -1,  1],
              [ 2,  2,  2],
              [-1, -1,  2]])
b = np.transpose(np.array([-1,  4, -5]))
# To solve the system we do

x = np.linalg.solve(A,b)
print("Real solution: ")
print(x)

Real solution: 
[ 1.  2. -1.]


**Using Jacobi Iteration:**

In [24]:
A = np.array([[ 2, -1,  1],
              [ 2,  2,  2],
              [-1, -1,  2]])

L = np.array([[ 0,  0,  0],
              [ 2,  0,  0],
              [-1, -1,  0]])

D = np.array([[ 2,  0,  0],
              [ 0,  2,  0],
              [ 0,  0,  2]])

U = np.array([[ 0, -1,  1],
              [ 0,  0,  2],
              [ 0,  0 , 0]])

D_inverse =  np. linalg. inv(D)
b = np.transpose(np.array([-1,  4, -5]))
D_inverse_b = np.dot(D_inverse, b)

BJ = np.dot(-D_inverse, L+U)
w,v=eig(BJ)

print('BJ evalues:', w)

x0 = np.transpose(np.array([ 0,  0,  0]))
x = list();
x.append(x0)

for i in range(10):
  x1 = np.dot(BJ, x0)+ D_inverse_b
  x0 = x1
  x.append(x1)
Aproximations = np.array(x)
print(Aproximations)

BJ evalues: [5.55111512e-17+1.11803399j 5.55111512e-17-1.11803399j
 2.60097238e-17+0.j        ]
[[ 0.          0.          0.        ]
 [-0.5         2.         -2.5       ]
 [ 1.75        5.         -1.75      ]
 [ 2.875       2.          0.875     ]
 [ 0.0625     -1.75       -0.0625    ]
 [-1.34375     2.         -3.34375   ]
 [ 2.171875    6.6875     -2.171875  ]
 [ 3.9296875   2.          1.9296875 ]
 [-0.46484375 -3.859375    0.46484375]
 [-2.66210938  2.         -4.66210938]
 [ 2.83105469  9.32421875 -2.83105469]]


*Because the spectral radius of matrix BJ,  P(BJ) = 1.12 > 1, Jacobi Iteration does not converge.*

**Using Gauss_Siedell Iteration:**

In [25]:
A = np.array([[ 2, -1,  1],
              [ 2,  2,  2],
              [-1, -1,  2]])

L = np.array([[ 0,  0,  0],
              [ 2,  0,  0],
              [-1, -1,  0]])

D = np.array([[ 2,  0,  0],
              [ 0,  2,  0],
              [ 0,  0,  2]])

U = np.array([[ 0, -1,  1],
              [ 0,  0,  2],
              [ 0,  0 , 0]])

DL_inverse =  np. linalg. inv(D+L)
b = np.transpose(np.array([-1,  4, -5]))
DL_inverse_b = np.dot(DL_inverse, b)

BGS = np.dot(-DL_inverse, U)
w,v=eig(BGS)

print('BGS evalues:', w)

x0 = np.transpose(np.array([ 0, 0, 0]))
x = list();
x.append(x0)

for i in range(10):
  x1 = np.dot(BGS, x0)+ DL_inverse_b
  x0 = x1
  x.append(x1)
Aproximations = np.array(x)
print(Aproximations)

BGS evalues: [ 0.  -0.5 -0.5]
[[ 0.          0.          0.        ]
 [-0.5         2.5        -1.5       ]
 [ 1.5         2.         -0.75      ]
 [ 0.875       1.875      -1.125     ]
 [ 1.          2.125      -0.9375    ]
 [ 1.03125     1.90625    -1.03125   ]
 [ 0.96875     2.0625     -0.984375  ]
 [ 1.0234375   1.9609375  -1.0078125 ]
 [ 0.984375    2.0234375  -0.99609375]
 [ 1.00976562  1.98632812 -1.00195312]
 [ 0.99414062  2.0078125  -0.99902344]]


*Because the spectral radius of matrix BGS,  P(BGS) = 0.5 < 1, Gauss-Siedell Iteration does converge.*

**Comparison of BJ vs BGS, P(BJ) = 1.12 < 0.05 = P(BGS), Jacobi Iteration does not converge at all while Gauss-Siedell Iteration does!**

# Problem 4

In [26]:
# To solve Ax=b, We start by constructing the arrays for A and b.

A = np.array([[ 1,  2, -2],
              [ 1,  1,  1],
              [ 2,  2,  1]])
b = np.transpose(np.array([ 7, 2, 5]))
# To solve the system we do

x = np.linalg.solve(A,b)
print("Real solution: ")
print(x)

Real solution: 
[ 1.  2. -1.]


**Using Jacobi Iteration:**

In [27]:
A = np.array([[1, 2, -2],
              [1, 1,  1],
              [2, 2, 1]])

L = np.array([[0, 0, 0],
              [1, 0, 0],
              [2, 2, 0]])

D = np.array([[1, 0, 0],
              [0, 1, 0],
              [0, 0, 1]])

U = np.array([[0, 2, -2],
              [0, 0,  1],
              [0, 0,  0]])

D_inverse =  np. linalg. inv(D)
b = np.transpose(np.array([ 7,  2,  5]))
D_inverse_b = np.dot(D_inverse, b)

BJ = np.dot(-D_inverse, L+U)
w,v=eig(BJ)

print('BJ evalues:', w)

x0 = np.transpose(np.array([ 0,  0,  0]))
x = list();
x.append(x0)

for i in range(6):
  x1 = np.dot(BJ, x0)+ D_inverse_b
  x0 = x1
  x.append(x1)
Aproximations = np.array(x)
print(Aproximations)

BJ evalues: [-5.40666289e-06+9.36461231e-06j -5.40666289e-06-9.36461231e-06j
  1.08133258e-05+0.00000000e+00j]
[[  0.   0.   0.]
 [  7.   2.   5.]
 [ 13. -10. -13.]
 [  1.   2.  -1.]
 [  1.   2.  -1.]
 [  1.   2.  -1.]
 [  1.   2.  -1.]]


*Because the spectral radius of matrix BJ,  P(BJ) = 1.08133258e-05 < 1, Jacobi Iteration does converge.*

**Using Gaus-Siedell iteration:**

In [28]:
L = np.array([[0, 0, 0],
              [1, 0, 0],
              [2, 2, 0]])

D = np.array([[1, 0, 0],
              [0, 1, 0],
              [0, 0, 1]])

U = np.array([[0, 2, -2],
              [0, 0,  1],
              [0, 0,  0]])

DL_inverse =  np. linalg. inv(D+L)
b = np.transpose(np.array([7,2,5]))
DL_inverse_b = np.dot(DL_inverse, b)

BGS = np.dot(-DL_inverse, U)
w,v=eig(BGS)

print('BGS evalues:', w)

x0 = np.transpose(np.array([ 0, 0, 0]))
x = list();
x.append(x0)

for i in range(5):
  x1 = np.dot(BGS, x0)+ DL_inverse_b
  x0 = x1
  x.append(x1)
Aproximations = np.array(x)
print(Aproximations)

BGS evalues: [0. 2. 2.]
[[   0.    0.    0.]
 [   7.   -5.    1.]
 [  19.  -18.    3.]
 [  49.  -50.    7.]
 [ 121. -126.   15.]
 [ 289. -302.   31.]]


*Because the spectral radius of matrix BGS,  P(BGS) = 2 > 1, Gauss-Siedell Iteration does not converge.*

**Comparison of BJ vs BGS, P(BJ) = 1.08133258e-05  < 2 = P(BGS), Jacobi Iteration does converge while Gauss-Siedell Iteration does not at all!**